data-prepper-plugins/otel-trace-raw-prepper/src/main/java/com/amazon/dataprepper/plugins/prepper/oteltrace/OTelTraceRawPrepper.java

```diff
@@ -94,11 +105,155 @@ public OTelTraceRawPrepper(final PluginSetting pluginSetting) {
+        if (rawSpan.getParentSpanId() == null || "".equals(rawSpan.getParentSpanId())) {
+            processRootRawSpan(rawSpan);
+        } else {
+            processDescendantRawSpan(rawSpan);
+        }
+    }
+
+    private void processRootRawSpan(final RawSpan rawSpan) {
+        final long now = System.currentTimeMillis();
+        final long nowPlusOffset = now + rootSpanFlushDelay;
+        final DelayedParentSpan delayedParentSpan = new DelayedParentSpan(rawSpan, nowPlusOffset);
+        delayedParentSpanQueue.add(delayedParentSpan);
+    }
+
+    private void processDescendantRawSpan(final RawSpan rawSpan) {
+        traceIdRawSpanSetMap.compute(rawSpan.getTraceId(), (traceId, rawSpanSet) -> {
+            if (rawSpanSet == null) {
+                rawSpanSet = new RawSpanSet();
+            }
+            rawSpanSet.addRawSpan(rawSpan);
+            return rawSpanSet;
+        });
+    }
+
+    private List<Record<String>> convertRawSpansToJsonRecords(final List<RawSpan> rawSpans) {
+        final List<Record<String>> records = new LinkedList<>();
+
+        for (RawSpan rawSpan : rawSpans) {
+            String rawSpanJson;
+            try {
+                rawSpanJson = rawSpan.toJson();
+            } catch (JsonProcessingException e) {
+                log.error("Unable to process invalid Span {}:", rawSpan, e);
+                spanErrorsCounter.increment();
+                totalProcessingErrorsCounter.increment();
+                continue;
+            }
+
+            records.add(new Record<>(rawSpanJson));
+        }
+
+        return records;
+    }
+
+    private List<RawSpan> getTracesToFlushByRootSpan() {
+        final List<RawSpan> recordsToFlush = new LinkedList<>();
+
+        for (DelayedParentSpan delayedParentSpan; (delayedParentSpan = delayedParentSpanQueue.poll()) != null;) {
+            RawSpan parentSpan = delayedParentSpan.getRawSpan();
+            recordsToFlush.add(parentSpan);
+
+            String traceGroup = parentSpan.getTraceGroup();
+            String parentSpanTraceId = parentSpan.getTraceId();
+
+            RawSpanSet rawSpanSet = traceIdRawSpanSetMap.get(parentSpanTraceId);
+            if (rawSpanSet != null) {
+                for (RawSpan rawSpan : rawSpanSet.getRawSpans()) {
+                    rawSpan.setTraceGroup(traceGroup);
+                    recordsToFlush.add(rawSpan);
+                }
+
+                traceIdRawSpanSetMap.remove(parentSpanTraceId);
+            }
+        }
+
+        return recordsToFlush;
+    }
+
+
+    private List<RawSpan> getTracesToFlushByGarbageCollection() {
+        final List<RawSpan> recordsToFlush = new LinkedList<>();
+
+        if (shouldGarbageCollect()) {
+            boolean isLockAcquired = gcLock.tryLock();
+
+            if (isLockAcquired) {
+                try {
+                    final long now = System.currentTimeMillis();
+                    lastTraceFlushTime = now;
+
+                    // fail-safe
```

Comment:
nit - unnecessary (or not descriptive enough). Could add docs at the method level to indicate what's going on.

Reply:
Will remove